magic_transit_ipsec_tunnel.go

```diff
@@ -0,0 +1,190 @@
+package cloudflare
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+)
+
+// Magic Transit IPsec Tunnel Error messages.
+const (
+	errMagicTransitIPsecTunnelNotModified = "When trying to modify IPsec tunnel, API returned modified: false"
+	errMagicTransitIPsecTunnelNotDeleted  = "When trying to delete IPsec tunnel, API returned deleted: false"
+)
+
+// MagicTransitIPsecTunnel contains information about an IPsec tunnel.
+type MagicTransitIPsecTunnel struct {
+	ID                 string     `json:"id,omitempty"`
+	CreatedOn          *time.Time `json:"created_on,omitempty"`
+	ModifiedOn         *time.Time `json:"modified_on,omitempty"`
+	Name               string     `json:"name"`
+	CustomerEndpoint   string     `json:"customer_endpoint"`
+	CloudflareEndpoint string     `json:"cloudflare_endpoint"`
+	InterfaceAddress   string     `json:"interface_address"`
+	Description        string     `json:"description,omitempty"`
+}
+
+// ListMagicTransitIPsecTunnelsResponse contains a response including IPsec tunnels.
+type ListMagicTransitIPsecTunnelsResponse struct {
+	Response
+	Result struct {
+		IPsecTunnels []MagicTransitIPsecTunnel `json:"ipsec_tunnels"`
+	} `json:"result"`
+}
+
+// GetMagicTransitIPsecTunnelResponse contains a response including zero or one IPsec tunnels.
+type GetMagicTransitIPsecTunnelResponse struct {
+	Response
+	Result struct {
+		IPsecTunnel MagicTransitIPsecTunnel `json:"ipsec_tunnel"`
+	} `json:"result"`
+}
+
+// CreateMagicTransitIPsecTunnelsRequest is an array of IPsec tunnels to create.
+type CreateMagicTransitIPsecTunnelsRequest struct {
+	IPsecTunnels []MagicTransitIPsecTunnel `json:"ipsec_tunnels"`
+}
+
+// UpdateMagicTransitIPsecTunnelResponse contains a response after updating an IPsec Tunnel.
+type UpdateMagicTransitIPsecTunnelResponse struct {
+	Response
+	Result struct {
+		Modified            bool                    `json:"modified"`
+		ModifiedIPsecTunnel MagicTransitIPsecTunnel `json:"modified_ipsec_tunnel"`
+	} `json:"result"`
+}
+
+// DeleteMagicTransitIPsecTunnelResponse contains a response after deleting an IPsec Tunnel.
+type DeleteMagicTransitIPsecTunnelResponse struct {
+	Response
+	Result struct {
+		Deleted            bool                    `json:"deleted"`
+		DeletedIPsecTunnel MagicTransitIPsecTunnel `json:"deleted_ipsec_tunnel"`
+	} `json:"result"`
+}
+
+// ListMagicTransitIPsecTunnels lists all IPsec tunnels for a given account
+//
+// API reference: https://api.cloudflare.com/#magic-ipsec-tunnels-list-ipsec-tunnels
+func (api *API) ListMagicTransitIPsecTunnels(ctx context.Context) ([]MagicTransitIPsecTunnel, error) {
+	if err := api.checkAccountID(); err != nil {
+		return []MagicTransitIPsecTunnel{}, err
+	}
+
+	uri := fmt.Sprintf("/accounts/%s/magic/ipsec_tunnels", api.AccountID)
```

Comment:
can we refactor this to instead accept the account ID as a parameter? api.AccountID isn’t concurrency safe and is causing issues with some Tenant support.

Reply:
Makes sense, thank you for raising. I have changed the signatures.